File: frontend/src/services/classroom.ts
// 课堂互动 API

import { api } from "./api";

const BASE = "/classroom";
const ADMIN_BASE = "/classroom/admin";

// ─── 类型 ───

export interface OptionItem {
  key: string;
  text: string;
}

export interface Activity {
  id: number;
  activity_type: "vote" | "fill_blank";
  title: string;
  options: OptionItem[] | null;
  correct_answer: string | null;
  allow_multiple: boolean;
  time_limit: number;
  status: "draft" | "active" | "ended";
  started_at: string | null;
  ended_at: string | null;
  created_by: number;
  created_at: string;
  response_count?: number;
  remaining_seconds?: number | null;
  my_answer?: string | null;
  stats?: ActivityStats | null;
  analysis_agent_id?: number | null;
  analysis_prompt?: string | null;
  analysis_status?: string | null;
  analysis_result?: string | null;
  analysis_context?: Record<string, any> | null;
  analysis_error?: string | null;
  analysis_updated_at?: string | null;
}

export interface ActivityStats {
  activity_id: number;
  total_responses: number;
  option_counts: Record<string, number> | null;
  correct_count: number;
  correct_rate: number | null;
  blank_slot_stats?: Array<{
    slot_index: number;
    correct_answer: string;
    total_count: number;
    correct_count: number;
    correct_rate: number | null;
    top_wrong_answers: Array<{ answer: string; count: number }>;
  }> | null;
  top_wrong_answers?: Array<{ answer: string; count: number }> | null;
}

export interface ActivityListResponse {
  items: Activity[];
  total: number;
  page: number;
  page_size: number;
}

export interface ActivityResult {
  id: number;
  activity_type: string;
  title: string;
  options: OptionItem[] | null;
  correct_answer: string | null;
  allow_multiple: boolean;
  my_answer: string | null;
  is_correct: boolean | null;
  stats: ActivityStats | null;
}

export interface ActivityCreateRequest {
  activity_type: "vote" | "fill_blank";
  title: string;
  options?: OptionItem[];
  correct_answer?: string;
  allow_multiple?: boolean;
  time_limit?: number;
  analysis_agent_id?: number;
  analysis_prompt?: string;
}

export interface ActivityEndRequest {
  analysis_agent_id?: number;
  analysis_prompt?: string;
}

export interface ActiveAgentOption {
  id: number;
  name: string;
}

// ─── API ───

export const classroomApi = {
  // 管理端
  create: async (data: ActivityCreateRequest): Promise<Activity> => {
    const resp = await api.post(ADMIN_BASE + "/", data);
    return resp.data as any;
  },
  update: async (id: number, data: Partial<ActivityCreateRequest>): Promise<Activity> => {
    const resp = await api.put(`${ADMIN_BASE}/${id}`, data);
    return resp.data as any;
  },
  remove: async (id: number): Promise<void> => {
    await api.delete(`${ADMIN_BASE}/${id}`);
  },

  bulkRemove: async (ids: number[]): Promise<{ deleted: number[]; skipped: number[] }> => {
    const resp = await api.post(`${ADMIN_BASE}/bulk-delete`, { ids });
    return resp.data as any;
  },

  duplicate: async (id: number): Promise<Activity> => {
    const resp = await api.post(`${ADMIN_BASE}/${id}/duplicate`);
    return resp.data as any;
  },
  restart: async (id: number): Promise<Activity> => {
    const resp = await api.post(`${ADMIN_BASE}/${id}/restart`);
    return resp.data as any;
  },
  list: async (params?: { skip?: number; limit?: number; status?: string }): Promise<ActivityListResponse> => {
    const resp = await api.get(ADMIN_BASE + "/", { params });
    return resp.data as any;
  },
  getDetail: async (id: number): Promise<Activity> => {
    const resp = await api.get(`${ADMIN_BASE}/${id}`);
    return resp.data as any;
  },
  start: async (id: number): Promise<Activity> => {
    const resp = await api.post(`${ADMIN_BASE}/${id}/start`);
    return resp.data as any;
  },
  end: async (id: number, data?: ActivityEndRequest): Promise<Activity> => {
    const resp = await api.post(`${ADMIN_BASE}/${id}/end`, data || {});
    return resp.data as any;
  },
  getStatistics: async (id: number): Promise<ActivityStats> => {
    const resp = await api.get(`${ADMIN_BASE}/${id}/statistics`);
    return resp.data as any;
  },
  getActiveAgents: async (): Promise<ActiveAgentOption[]> => {
    const resp = await api.get("/ai-agents/active");
    const rows = Array.isArray(resp.data) ? resp.data : [];
    return rows.map((row: any) => ({
      id: Number(row?.id || 0),
      name: String(row?.name || row?.agent_name || `智能体${row?.id ?? ""}`),
    })).filter((row: ActiveAgentOption) => Number.isFinite(row.id) && row.id > 0);
  },

  // 学生端
  getActive: async (): Promise<Activity[]> => {
    const resp = await api.get(BASE + "/active");
    return resp.data as any;
  },
  getActivity: async (id: number): Promise<Activity> => {
    const resp = await api.get(`${BASE}/${id}`);
    return resp.data as any;
  },
  respond: async (id: number, answer: string): Promise<{ id: number; answer: string; is_correct: boolean | null; submitted_at: string }> => {
    const resp = await api.post(`${BASE}/${id}/respond`, { answer });
    return resp.data as any;
  },
  getResult: async (id: number): Promise<ActivityResult> => {
    const resp = await api.get(`${BASE}/${id}/result`);
    return resp.data as any;
  },
};

File: backend/app/schemas/classroom.py
"""课堂互动 Pydantic schemas"""

from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Literal
from datetime import datetime


ActivityType = Literal["vote", "fill_blank"]
# 说明：产品文档曾提及 5 种活动类型，但当前仅 vote(投票)/fill_blank(填空) 两种已实现，
# 前端也仅消费这两种。其余类型未实现，强行扩展会引入空壳端点，故保持现状。


class OptionItem(BaseModel):
    key: str = Field(..., description="选项标识 A/B/C/D")
    text: str = Field(..., description="选项内容")


class ActivityCreate(BaseModel):
    activity_type: ActivityType
    title: str = Field(..., min_length=1, max_length=200)
    # 班级名称：留空的旧活动不会向学生端开放。
    class_name: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = None
    options: Optional[List[OptionItem]] = None
    correct_answer: Optional[str] = None
    allow_multiple: bool = False
    time_limit: int = Field(60, ge=0, description="秒, 0=无限制")
    analysis_agent_id: Optional[int] = None
    analysis_prompt: Optional[str] = None


class ActivityUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    class_name: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = None
    options: Optional[List[OptionItem]] = None
    correct_answer: Optional[str] = None
    allow_multiple: Optional[bool] = None
    time_limit: Optional[int] = Field(None, ge=0)
    analysis_agent_id: Optional[int] = None
    analysis_prompt: Optional[str] = None


class ActivityEndRequest(BaseModel):
    analysis_agent_id: Optional[int] = None
    analysis_prompt: Optional[str] = None


class ActivityResponse(BaseModel):
    id: int
    activity_type: str
    title: str
    class_name: Optional[str] = None
    description: Optional[str] = None
    options: Optional[list] = None
    correct_answer: Optional[str] = None
    allow_multiple: bool
    time_limit: int
    status: str
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None
    created_by: int
    created_at: datetime
    response_count: int = 0
    remaining_seconds: Optional[int] = None
    analysis_agent_id: Optional[int] = None
    analysis_prompt: Optional[str] = None
    analysis_status: Optional[str] = None
    analysis_result: Optional[str] = None
    analysis_context: Optional[dict] = None
    analysis_error: Optional[str] = None
    analysis_updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class ActivityStudentView(BaseModel):
    """学生视角 — 隐藏正确答案"""
    id: int
    activity_type: str
    title: str
    class_name: Optional[str] = None
    description: Optional[str] = None
    options: Optional[list] = None
    allow_multiple: bool
    time_limit: int
    status: str
    started_at: Optional[datetime] = None
    remaining_seconds: Optional[int] = None
    my_answer: Optional[str] = None


class ResponseSubmit(BaseModel):
    answer: str = Field(..., min_length=1, max_length=500)


class ResponseResult(BaseModel):
    id: int
    answer: str
    is_correct: Optional[bool] = None
    submitted_at: datetime


class ActivityStats(BaseModel):
    activity_id: int
    total_responses: int
    option_counts: Optional[dict] = None
    correct_count: int = 0
    correct_rate: Optional[float] = None
    blank_slot_stats: Optional[list] = None
    top_wrong_answers: Optional[list] = None


class ActivityListResponse(BaseModel):
    items: list
    total: int
    page: int
    page_size: int


class ActivityResultView(BaseModel):
    """活动结束后的结果视图"""
    id: int
    activity_type: str
    title: str
    options: Optional[list] = None
    allow_multiple: bool
    my_answer: Optional[str] = None
    is_correct: Optional[bool] = None
    stats: Optional[ActivityStats] = None
